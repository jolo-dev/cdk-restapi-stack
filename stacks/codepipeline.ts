import { Repository } from '@aws-cdk/aws-codecommit';
import { Pipeline } from '@aws-cdk/aws-codepipeline';
import { CodeBuildProject } from '@aws-cdk/aws-events-targets';
import { Stack, Construct, StackProps } from '@aws-cdk/core';
import { BuildStage } from '../src/BuildStage';
import { SourceStage } from '../src/SourceStage';

export class FourDPipelineStack extends Stack {

  private readonly repositoryArn: string;

  constructor(scope: Construct, id: string, repositoryArn: string = '4dt-api-node', props?: StackProps) {
    super(scope, id, props);

    this.repositoryArn = repositoryArn;
    const codepipeline = new Pipeline(this, `${id}Pipeline`);
    const repository = Repository.fromRepositoryName(this, `${id}-repository`, this.repositoryArn);

    // CodeCommit
    const source = new SourceStage(repository);
    codepipeline.addStage({
      stageName: 'Source',
      actions: [source.getCodeCommitSourceAction()],
    });

    // Build
    const build = new BuildStage(this, 'build',
      {
        install: {
          commands: ['npm install -g pnpm', 'pnpm install'],
        },
        build: {
          commands: ['pnpm build'],
        },
      });

    const buildAction = build.buildAction(source.getSourceOutput(), 'build');

    // Deploy
    const cdkDeploy = new BuildStage(this, 'deploy',
      {
        post_build: {
          commands: ['pnpm deploy'],
        },
      });
    const cdkDeployAction = cdkDeploy.buildAction(source.getSourceOutput(), 'deploy');

    codepipeline.addStage({
      stageName: 'ProjenBuild', // projens build includes testing and linting
      actions: [buildAction, cdkDeployAction],
    });

    repository.onCommit('OnCommit', {
      target: new CodeBuildProject(build.getPipelineProject()),
    });
  }
}