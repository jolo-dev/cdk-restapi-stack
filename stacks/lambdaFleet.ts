import { Stack, Construct, StackProps } from '@aws-cdk/core';
import { Code, Function, Runtime, Tracing } from '@aws-cdk/aws-lambda';
import fs from 'fs'
import path from 'path'

export class LambdaFleetStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps){
        super(scope, id, props);

        const buildPath = '../lambdas/build'
        const lambdas = fs.readdirSync(path.dirname(buildPath))
        lambdas.forEach(lambda => {
            const lambdaName = lambda.replace('.js', '')
            new Function(this, `${lambdaName}Function`, {
                runtime: Runtime.NODEJS_14_X,
                handler: 'app.handler',
                code: Code.fromAsset(`buildPath/${lambda}`),
                tracing: Tracing.ACTIVE,
              })
        })
    }
}