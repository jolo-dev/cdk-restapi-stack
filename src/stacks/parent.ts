import { Construct, Stack, StackProps } from '@aws-cdk/core';
import { FourDPipelineStack } from './codepipeline';
import { LambdaFleetStack } from './lambdaFleet';


export class CdkParentStack extends Stack {
  constructor(scope: Construct, id: string, repositoryName: string, props?: StackProps) {
    super(scope, id, props);

    new FourDPipelineStack(this, 'FourDPipelineStack', repositoryName);
    new LambdaFleetStack(this, 'LambdaFleetStack', 'lambdas');
  }
}