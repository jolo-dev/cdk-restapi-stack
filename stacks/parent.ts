import { Construct, Stack, StackProps } from '@aws-cdk/core';
import { FourDPipelineStack } from './codepipeline';
import { LambdaFleetStack } from './lambdaFleet';

const repositoryName = process.env.GIT_REPOSITORY ?? '4dt-api-node';
export class CdkParentStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new FourDPipelineStack(this, 'FourDPipelineStack', repositoryName);
    new LambdaFleetStack(this, 'LambdaFleetStack', 'lambdas');
  }
}