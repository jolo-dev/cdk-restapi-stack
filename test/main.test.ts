import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { CdkParentStack } from '../src/stacks/parent';

const env = {
  account: process.env.CDK_DEFAULT_ACCOUNT ?? '123456789010',
  region: 'eu-west-1',
};
test('InfraStack', () => {
  const app = new App({
    context: {
      vpcId: 'vpc-123456789',
      privateSubnet1: 'private-subnet-1',
      privateSubnet2: 'private-subnet-2',
    },
  });


  const stack = new CdkParentStack(app, 'test-stack', 'test-repo', { env });
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});