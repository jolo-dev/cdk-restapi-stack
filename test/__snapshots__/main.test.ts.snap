// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfraStack 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615ArtifactHashD944D272": Object {
      "Description": "Artifact hash for asset \\"56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615\\"",
      "Type": "String",
    },
    "AssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615S3BucketBC8F7C7A": Object {
      "Description": "S3 bucket for asset \\"56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615\\"",
      "Type": "String",
    },
    "AssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615S3VersionKey4A69B7BE": Object {
      "Description": "S3 key for asset version \\"56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615\\"",
      "Type": "String",
    },
    "AssetParametersb018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5caArtifactHashFB552854": Object {
      "Description": "Artifact hash for asset \\"b018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5ca\\"",
      "Type": "String",
    },
    "AssetParametersb018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5caS3Bucket19AC43B7": Object {
      "Description": "S3 bucket for asset \\"b018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5ca\\"",
      "Type": "String",
    },
    "AssetParametersb018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5caS3VersionKey22A0DCF5": Object {
      "Description": "S3 key for asset version \\"b018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5ca\\"",
      "Type": "String",
    },
    "AssetParametersff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6ArtifactHash83014A1D": Object {
      "Description": "Artifact hash for asset \\"ff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6\\"",
      "Type": "String",
    },
    "AssetParametersff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6S3BucketBD08C9D3": Object {
      "Description": "S3 bucket for asset \\"ff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6\\"",
      "Type": "String",
    },
    "AssetParametersff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6S3VersionKey114817F5": Object {
      "Description": "S3 key for asset version \\"ff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "FourDPipelineStackNestedStackFourDPipelineStackNestedStackResourceCC0A20CE": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.eu-west-1.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParametersb018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5caS3Bucket19AC43B7",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersb018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5caS3VersionKey22A0DCF5",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersb018189d0aa33ed6ed6e4851dcf06c32d460f85d7ce6688c77987352a4afa5caS3VersionKey22A0DCF5",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaFleetStackNestedStackLambdaFleetStackNestedStackResource438355F9": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "referencetoteststackAssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615S3BucketFD256717Ref": Object {
            "Ref": "AssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615S3BucketBC8F7C7A",
          },
          "referencetoteststackAssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615S3VersionKey41992936Ref": Object {
            "Ref": "AssetParameters56ff781865cb28e850ba89eb90a0aa9b48f31ad4ff3bf2231db34f87cbb01615S3VersionKey4A69B7BE",
          },
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.eu-west-1.",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParametersff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6S3BucketBD08C9D3",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6S3VersionKey114817F5",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersff86a9ccac1270080e7b2640ec2bb1006c76c99b6efdf23f3293931f053e93f6S3VersionKey114817F5",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
